ori/volume/server/volume_create.go

```diff
@@ -17,260 +17,92 @@ package server
-	if volumeSizeBytes < snapshot.Status.RestoreSize.Value() {
-		log.Info(fmt.Sprintf("Requested volume size %d is less than the size %d for the source snapshot", volumeSizeBytes, snapshot.Status.RestoreSize.Value()), "snapshotName", snapshot.Name)
-		//r.Eventf(volume, corev1.EventTypeWarning, ReasonVolumeSizeToSmall, "Requested volume size %d is less than the size %d for the source snapshot %s", volumeSizeBytes, snapshot.Status.RestoreSize.Value(), snapshot.Name)
-	}
-
-	return nil
-}
-
-func (s *Server) createSnapshot(ctx context.Context, log logr.Logger, volume *AggregateCephVolume) error {
-	if err := s.client.Create(ctx, volume.ImagePvc); err != nil {
-		return fmt.Errorf("error creating pvc: %w", err)
-	}
-	log.V(1).Info(fmt.Sprintf("created image pvc %s", client.ObjectKeyFromObject(volume.ImagePvc)))
-
-	snapshot := &snapshotv1.VolumeSnapshot{
-		TypeMeta: metav1.TypeMeta{
-			Kind:       "VolumeSnapshot",
-			APIVersion: snapshotv1.SchemeGroupVersion.String(),
-		},
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      volume.ImagePvc.Name,
-			Namespace: volume.ImagePvc.Namespace,
-		},
-		Spec: snapshotv1.VolumeSnapshotSpec{
-			Source: snapshotv1.VolumeSnapshotSource{
-				PersistentVolumeClaimName: &volume.ImagePvc.Name,
-			},
-			VolumeSnapshotClassName: pointer.String(s.rookPoolStorageClass),
 		},
-	}
-	metautils.SetLabel(snapshot, volumev1alpha1.ManagerLabel, volumev1alpha1.VolumeCephletManager)
-
-	if err := s.client.Create(ctx, snapshot); err != nil {
-		return fmt.Errorf("error creating snapshot: %w", err)
-	}
-	log.V(1).Info(fmt.Sprintf("created snapshot %s for image pvc %s", client.ObjectKeyFromObject(snapshot), client.ObjectKeyFromObject(volume.ImagePvc)))
+	}, nil
 
-	return nil
 }
 
-func (s *Server) createVolume(ctx context.Context, log logr.Logger, volume *AggregateCephVolume) (retErr error) {
-
-	if err := s.handleImagePopulation(ctx, log, volume); err != nil {
-		return fmt.Errorf("error populate image: %w", err)
-	}
-
-	c, cleanup := s.setupCleaner(ctx, log, &retErr)
-	defer cleanup()
-
-	log.V(1).Info("Creating pvc")
-	if err := s.client.Create(ctx, volume.Pvc); err != nil {
-		return fmt.Errorf("error creating pvc: %w", err)
-	}
-	c.Add(func(ctx context.Context) error {
-		if err := s.client.Delete(ctx, volume.Pvc); client.IgnoreNotFound(err) != nil {
-			return fmt.Errorf("error deleting pvc: %w", err)
-		}
-		return nil
-	})
-
-	log.V(1).Info("Patching pvc as created")
-	if err := apiutils.PatchCreated(ctx, s.client, volume.Pvc); err != nil {
-		return fmt.Errorf("error patching pvc as created: %w", err)
+func (s *Server) createImage(ctx context.Context, log logr.Logger, cfg *CephVolumeConfig) (retErr error) {
+	image, err := s.imageStore.Create(ctx, cfg.image)
+	if err != nil {
+		return fmt.Errorf("failed to create image: %w", err)
 	}
 
-	// Reset cleaner since everything from now on operates on a consistent volume
-	c.Reset()
+	cfg.image = image
```

Comment:
Don't mutate the `cfg.image` here but instead return an `AggregatedImage` (if that type doesn't exist, create it).
Idea behind is the following: For creating, you first calculate all necessary artefacts (which results in a `CephVolumeConfig`). This is the parameter for the `createImage` call which produces an `AggregatedImage` (internal image with additional info required for ori). This way, there's always a clear separation between artefacts required for creation and the result of creation / getting existing images.